modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java

```diff
@@ -0,0 +1,303 @@
+
+            startCluster(cluster);
+        }
+
+        PartialNode someNode = cluster.values().iterator().next();
+
+        RaftGroupService raftClient = someNode.raftClient;
+
+        assertThat(raftClient.refreshLeader(), willCompleteSuccessfully());
+
+        long firstSafeTime = calculateSafeTime(someNode.clock);
+
+        // Send command with safe time X.
+        sendSafeTimeSyncCommand(raftClient, firstSafeTime, false);
+
+        // Stop all nodes
+        for (PartialNode node : cluster.values()
+        ) {
+            node.stop();
+        }
+
+        // And restart.
+        startCluster(cluster);
+
+        // Send command with safe time less than previously applied to the leader before the restart
+        // and verify that SafeTimeReorderException is thrown.
+        sendSafeTimeSyncCommand(someNode.raftClient, firstSafeTime - 1, true);
+
+        sendSafeTimeSyncCommand(someNode.raftClient, calculateSafeTime(someNode.clock), false);
+    }
+
+    private void startCluster(Map<String, PartialNode> cluster) throws Exception {
+        Collection<CompletableFuture<Void>> startingFutures = new ArrayList<>(cluster.size());
+        for (PartialNode node : cluster.values()) {
+            startingFutures.add(node.start());
+        }
+
+        CompletableFuture<Void> clusterReadyFuture = CompletableFuture.allOf(startingFutures.toArray(CompletableFuture[]::new));
+
+        assertThat(clusterReadyFuture, willCompleteSuccessfully());
+    }
+
+    private static void sendSafeTimeSyncCommand(
+            RaftGroupService raftClient,
+            long safeTime,
+            boolean expectSafeTimeReorderException
+    ) {
+        CompletableFuture<Object> safeTimeCommandFuture = raftClient.run(
+                REPLICA_MESSAGES_FACTORY
+                        .safeTimeSyncCommand()
+                        .safeTimeLong(safeTime)
+                        .build()
+        );
+
+        if (expectSafeTimeReorderException) {
+            assertThat(safeTimeCommandFuture, CompletableFutureExceptionMatcher.willThrow(SafeTimeReorderException.class));
+        } else {
+            assertThat(safeTimeCommandFuture, willCompleteSuccessfully());
+        }
+    }
+
+    private static long calculateSafeTime(HybridClock clock) {
+        return clock.now().addPhysicalTime(CLOCK_SKEW).longValue();
+    }
+
+    private class PartialNode {
+        private final String nodeName;
+
+        private final HybridClock clock;
+
+        private ClusterService clusterService;
+
+        private Loza raftManager;
+
+        private RaftGroupService raftClient;
+
+        PartialNode(String nodeName) {
+            this.nodeName = nodeName;
+            this.clock = new HybridClockImpl();
+        }
+
+        CompletableFuture<Void> start() throws Exception {
+            clusterService = ClusterServiceTestUtils.clusterService(nodeName, port++, NODE_FINDER);
```

Comment:
There's a chance that you wanted to call `start()` in parallel. If that's true then `port++` is unsafe

Reply:
Yep, you are right. Reworked to AtomicInteger.